### ISSUE TITLE
USR1 signal spawns additional refresh loop and doesn't clean up existing

### ISSUE THREAD

**Issue by mattlqx** [coredns/coredns#2396]
I'm using 1.2.6 and trying to figure out why I keep hitting Route 53 API throttling limits as part of #2353 . I adjusted the refresh loop to 5 minutes and upped the items per page to 1000 and reduced the number of daemons using the route53 plugin to just two (with 4 zones configured). Still hitting limits after a while of running.

Here's a diff of my only changes to the plugin:

```diff
diff --git a/plugin/route53/route53.go b/plugin/route53/route53.go
index 7420e267..3345c50f 100644
--- a/plugin/route53/route53.go
+++ b/plugin/route53/route53.go
@@ -84,7 +84,7 @@ func (h *Route53) Run(ctx context.Context) error {
                        case <-ctx.Done():
                                log.Infof("Breaking out of Route53 update loop: %v", ctx.Err())
                                return
-                       case <-time.After(1 * time.Minute):
+                       case <-time.After(5 * time.Minute):
                                if err := h.updateZones(ctx); err != nil && ctx.Err() == nil /* Don't log error if ctx expired. */ {
                                        log.Errorf("Failed to update zones: %v", err)
                                }
@@ -173,6 +173,7 @@ func (h *Route53) updateZones(ctx context.Context) error {
                                newZ.Upstream = *h.upstream
                                in := &route53.ListResourceRecordSetsInput{
                                        HostedZoneId: aws.String(hostedZone.id),
+                                       MaxItems: aws.String("1000"),
                                }
                                err = h.client.ListResourceRecordSetsPagesWithContext(ctx, in,
                                        func(out *route53.ListResourceRecordSetsOutput, last bool) bool {
```

I'm sending my daemons periodic SIGUSR1 to refresh the credentials file (since the credential is from an instance profile, it uses a session token that changes semi-frequently). My hunch was that for every USR1, an additional zone refresh loop would be created and they'd stack up and eventually hammer the AWS API reaching rate limits.

So I compiled coredns with debug symbols and whipped out gdb. Here's what I found. https://gist.github.com/mattlqx/0958888a2845757f5d41faaed3d3f556

The loop was set for 5 minutes. I sent the first USR1 approx. 2 minutes after the first loop firing that off. First loop fires 3 minutes after that. And so on. To summarize it:

```
Fri Dec 14 17:15:23 UTC 2018 Loop 1 refreshes
Fri Dec 14 17:17:10 UTC 2018 Loop 2 refreshes
Fri Dec 14 17:20:23 UTC 2018 Loop 1 refreshes
Fri Dec 14 17:22:11 UTC 2018 Loop 2 refreshes
Fri Dec 14 17:23:12 UTC 2018 USR1 sent, Loop 3 load
Fri Dec 14 17:25:24 UTC 2018 Loop 1 refreshes
Fri Dec 14 17:27:11 UTC 2018 Loop 2 refreshes
Fri Dec 14 17:28:14 UTC 2018 Loop 3 refreshes
```

Hope this is thorough enough! I don't know enough about Go to really know where to fix from here.

**Comment by mattlqx** [comment]
Looks like a manifestation of https://github.com/mholt/caddy/issues/2358

**Comment by miekg** [comment]
yes that would be a good one to fix - maybe good-first-issue?

**Comment by yongtang** [comment]
@mattlqx Would appreciate if you could send the PR.

**Comment by mattlqx** [comment]
This is totally out of my wheelhouse unfortunately. I actually know very little about Go.

**Comment by fturib** [comment]
@yuewko proposed a PR to fix the issue in Caddy. Thanks Ko !

**Comment by miekg** [comment]
we should have updated to that new version.

**Comment by mattlqx** [comment]
@miekg This is unfortunately still happening in CoreDNS 1.6.9. I finally switched from restarting to reloading on my infrastructure and within a day was hitting AWS Route 53 API ratelimiting again. I did the same debug test that I did when opening this issue on the currrent release and can see additional updateZone loops every time I send SIGUSR1 to the process. Should this be re-opened or should I file a whole new issue?
